homeassistant/components/sensor/__init__.py

```diff
@@ -141,7 +149,100 @@ def capability_attributes(self) -> Mapping[str, Any] | None:
+        attrs: dict[str, Any] = {}
         if last_reset := self.last_reset:
-            return {ATTR_LAST_RESET: last_reset.isoformat()}
+            attrs[ATTR_LAST_RESET] = last_reset.isoformat()
+        if native_value := self.native_value:
+            attrs[ATTR_NATIVE_VALUE] = native_value
+
+        return attrs
 
+    @property
+    def native_value(self) -> StateType:
+        """Return the value reported by the sensor."""
+        return self._attr_native_value
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the sensor, if any."""
+        if hasattr(self, "_attr_native_unit_of_measurement"):
+            return self._attr_native_unit_of_measurement
+        if hasattr(self, "entity_description"):
+            return self.entity_description.native_unit_of_measurement
         return None
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the entity, after unit conversion."""
+        if (
+            hasattr(self, "_attr_unit_of_measurement")
+            and self._attr_unit_of_measurement is not None
+        ):
+            return self._attr_unit_of_measurement
+        if (
+            hasattr(self, "entity_description")
+            and self.entity_description.unit_of_measurement is not None
+        ):
+            return self.entity_description.unit_of_measurement
+
+        native_unit_of_measurement = self.native_unit_of_measurement
+
+        if native_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            return self.hass.config.units.temperature_unit
+
+        return native_unit_of_measurement
+
+    @property
+    def state(self) -> Any:
+        """Return the state of the sensor and perform unit conversions, if needed."""
+        # Test if _attr_state has been set in this instance
+        if "_attr_state" in self.__dict__:
+            return self._attr_state
+
+        unit_of_measurement = self.native_unit_of_measurement
+        value = self.native_value
+
+        units = self.hass.config.units
+        if (
+            value is not None
+            and unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT)
+            and unit_of_measurement != units.temperature_unit
+        ):
+            if (
+                self.device_class != DEVICE_CLASS_TEMPERATURE
+                and not self._temperature_conversion_reported
+            ):
+                self._temperature_conversion_reported = True
+                report_issue = self._suggest_report_issue()
+                _LOGGER.warning(
+                    "Entity %s (%s) with device_class %s reports a temperature in "
+                    "%s which will be converted to %s, this is deprecated and will"
+                    " be removed from Home Assistant Core 2022.2. Please update "
+                    "your configuration if device_class is manually configured, "
+                    "otherwise %s",
```

Comment:
```suggestion
                    "Entity %s (%s) with device_class %s reports a temperature in "
                    "%s which will be converted to %s. Temperature conversion for "
                    "entities without correct device_class is deprecated and will"
                    " be removed from Home Assistant Core 2022.2. Please update "
                    "your configuration if device_class is manually configured, "
                    "otherwise %s",
```